### PR TITLE
[DDW-700] Share ICO public key

### PR DESCRIPTION
**Pull request by danielmain**
This PR Add the possibility to display the IO public key.

https://jira.iohk.io/browse/DDW-700

## Todos

- [x] Refactoring current public key component and dialog
- [x] Add new api call postAccountKey
- [x] Make the created ICO Public key available in wallet store
- [x] Add new stings for en-US and ja-JP

## Screenshots

- Screenshot
image
image
image
image
image
image

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/messages/GGKFXSKC6)
- [ ] Test

---

## Test Cases

Scenario 1: View wallet ICO public key field
Given I am on wallet Settings page
Then I can view wallet ICO public key field
And the ICO public key is not displayed

Scenario 2: View ICO public key
Given I am on wallet Settings page
And wallet ICO public key field is displayed
When I select “Reveal” button
Then my ICO wallet public key is shown

Scenario 3: Copying the wallet ICO public key
Given that I have revealed my wallet ICO public key
When I select the copy icon
Then wallet ICO public key is copied onto the clipboard

Scenario 4: Hiding the wallet ICO public key
Given that I have revealed my ICO wallet public key
When I select “Hide” button
Then wallet ICO public key is hidden

Scenario 5: Show QR Code
Given that I have revealed my wallet ICO public key
When I select “QR” button
Then a modal dialog opens
And QR Code for the wallet ICO public key is displayed
When I scan the QR Code
Then the correct wallet ICO public key is displayed
When I select ‘Close’
The modal dialog box is dismissed

Scenario 6: Ensure text is correct
Given I am on wallet Settings page
And I am on ‘lang’ language
Then the followings are correct

Translations and screens check the JIRA Issue please https://jira.iohk.io/browse/DDW-700

### Basics

- [x] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [x] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [x] PR has a good description that summarizes all changes
- [x] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [x] In English
  - [x] In Japanese
- [x] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [x] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [x] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [x] There are no *flow* errors or warnings (`yarn flow:test`)
- [x] There are no *lint* errors or warnings (`yarn lint`)
- [x] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [x] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [x] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [x] Important parts of the code are properly commented and documented
- [x] Code is properly typed with flow
- [x] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Comment by gabriela-ponce** [comment]
@danielmain Here's a list of some issues on build 18403:

- When you insert a short incorrect password to reveal the key, you get the message "An Error Occurred" instead of "Incorrect Password. It doesn't happen if the incorrect password is over 10 characters. Check [this](https://drive.google.com/file/d/1VPaBagQYMYPgtK0Pip_6AiC826dBvKiO/view?usp=sharing) video. A similar issue was addressed before for the public key on [this](https://github.com/input-output-hk/daedalus/pull/2580) PR (mentioning just in case the fix is relevant).
- When you switch between wallets, the ICO key can be revealed without inserting the password again. For the public key, you need to insert the password each time you switch to a different wallet, so I think both of these should work the same way. Check [this](https://drive.google.com/file/d/1Dk7HS6zK-eJzFZfzJgcQoSVlHOpkZWqz/view?usp=sharing) video.
- When you reveal the ICO key on any wallet, if you switch to another wallet, no password is asked to reveal the ICO key. Also, the key displayed is from the first wallet. Check [this](https://drive.google.com/file/d/10fHvNJWZECZUIugmASyPIa1HNXATUfEi/view?usp=sharing) video.
- (Related to the previous issue) When you reveal the ICO key on a wallet and switch to another and try to copy it, you get a console error and no notification displayed. Check [this](https://drive.google.com/file/d/1O0J_D6cZlK_Q_pZqrD6DGDUbg4Mjng3q/view?usp=sharing) video.

**Comment by danielmain** [comment]
@gabriela-ponce thank you for the detailed testing and including videos! I fixed all the Issues you found and tested on mac and additionally on Linux as well. Would appreciate a retesting.

File: source/renderer/app/api/api.js [hunk 1]
```diff
@@ -41,6 +41,8 @@ import { deleteLegacyTransaction } from './transactions/requests/deleteLegacyTra
 import { selectCoins } from './transactions/requests/selectCoins';
 import { createExternalTransaction } from './transactions/requests/createExternalTransaction';
 import { getPublicKey } from './transactions/requests/getPublicKey';
+import { getICOPublicKey } from './transactions/requests/getICOPublicKey';
+import type { ICOPublicKeyParams } from './transactions/requests/getICOPublicKey';
```
Comment (daniloprates) [inline source/renderer/app/api/api.js:45]: Please, move it to the bottom, along with the other transaction types (line 144)

File: source/renderer/app/api/api.js [hunk 2]
```diff
@@ -1272,6 +1274,24 @@ export default class AdaApi {
     }
   };
 
+  getICOPublicKey = async (
+    request: ICOPublicKeyParams
+  ): Promise<string> | null => {
```
Comment (daniloprates) [inline source/renderer/app/api/api.js:1279]: Do we need `null` here?

File: source/renderer/app/api/api.js (hunk 1 above)
Comment (daniloprates) [reply source/renderer/app/api/api.js:45]: Also, please move this type into `api/transactions/types.js`

File: source/renderer/app/components/wallet/settings/GenericPublicKeyField.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
+// @flow
```
Comment (daniloprates) [inline source/renderer/app/components/wallet/settings/GenericPublicKeyField.js:1]: Not sure it needs the `Generic` in the name. Consider removing it

File: source/renderer/app/components/wallet/settings/GenericPublicKeyField.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import { LOCALES } from '../../../../../common/types/locales.types';
+import styles from './GenericPublicKeyField.scss';
+
+type Props = {
+  publicKey: string,
+  locale: Locale,
+  onCopyPublicKey: Function,
+  onShowQRCode: Function,
+  onOpenWalletKeyDialog: Function,
+  t: Function,
```
Comment (daniloprates) [inline source/renderer/app/components/wallet/settings/GenericPublicKeyField.js:24]: Please use context instead of props for `intl`. Check [this file](https://github.com/input-output-hk/daedalus/blob/develop/source/renderer/app/components/assets/AssetContent.js#L228).

Also, it's nice to be more descriptive in the prop names

**Review by daniloprates — CHANGES_REQUESTED** [review]
Amazing job, @danielmain 🚀 

Please, check some comments I've left.

File: package.json [hunk 5]
```diff
@@ -224,7 +224,8 @@
     "react-custom-scrollbars": "4.2.1",
     "react-datetime": "3.0.4",
     "react-dom": "16.13.1",
-    "react-intl": "2.7.2",
+    "react-intl": "2.9.0",
+    "react-intl-hooks": "^1.0.11",
```
Comment (DominikGuzei) [inline package.json:228]: In general we avoid updating dependencies within feature PRs if not necessary - we don't know if this change might cause unrelated trouble somewhere in the app. If this change is not required, please move it to its own PR which can be tested and approved in isolation.

File: source/renderer/app/components/wallet/settings/ICOPublicKeyDialog.js [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+type Props = {
+  onRevealPublicKey: Function,
+  onClose: Function,
+  error: ?LocalizableError,
+  hasReceivedICOPublicKey: boolean,
+};
+
+@observer
+export default class ICOPublicKeyDialog extends Component<Props> {
```
Comment (DominikGuzei) [inline source/renderer/app/components/wallet/settings/ICOPublicKeyDialog.js:44]: Let's also use a functional component here?

File: source/renderer/app/components/wallet/settings/PublicKeyQRCodeDialog.js [hunk 7]
```diff
@@ -1,38 +1,34 @@
...
-    defaultMessage: '!!!Wallet Public Key',
-    description: 'Title for the "Wallet Public Key QR Code" dialog.',
-  },
-  copyPublicKeyLabel: {
-    id: 'wallet.settings.copyPublicKey',
-    defaultMessage: '!!!Copy public key',
-    description: 'Copy public key label.',
-  },
-});
+export type $npm$ReactIntl$MessageDescriptor = {
```
Comment (DominikGuzei) [inline source/renderer/app/components/wallet/settings/PublicKeyQRCodeDialog.js:15]: Why such a cryptic type name? And why is it exported from this component if it's a generic type?

**Review by DominikGuzei — CHANGES_REQUESTED** [review]

**Comment by daniloprates** [comment]
@danielmain, I was wondering if we really need to update `intl` and add `react-intl-hooks`. Whenever possible, we have dedicated tasks for this kind of update. Please, check the [AssetContent.js](https://github.com/input-output-hk/daedalus/blob/develop/source/renderer/app/components/assets/AssetContent.js#L77) file to see if you can have this implement the same way. If so, please revert the packages changes.

**Comment by daniloprates** [comment]
@danielmain one other thing I've noticed, when you enter the spending password, you need to click again the eye icon to see the ID. It should be displayed right after the password is entered.

File: source/renderer/app/api/api.js (hunk 2 above)
Comment (danielmain) [reply source/renderer/app/api/api.js:1279]: What do you like more? 'any' or 'string | null' ?

File: source/renderer/app/components/wallet/settings/GenericPublicKeyField.js (hunk 4 above)
Comment (danielmain) [reply source/renderer/app/components/wallet/settings/GenericPublicKeyField.js]: Since `t()` is wide used for translations (i18next) I just used it. But of course I will change it to intl()

File: package.json (hunk 5 above)
Comment (danielmain) [reply package.json]: Thank you for the tip @DominikGuzei ! I will use injectIntl. No need for extra dependencies

File: source/renderer/app/components/wallet/settings/ICOPublicKeyDialog.js (hunk 6 above)
Comment (danielmain) [reply source/renderer/app/components/wallet/settings/ICOPublicKeyDialog.js:43]: I tried really but I couldn't make it working with MobxReactForm

**Comment by gabriela-ponce** [comment]
@danielmain The fixes for [these](https://github.com/input-output-hk/daedalus/pull/2604#issuecomment-867876889) issues look good, great work 👍 
I found another small inconsistency between the "[Reveal public key](https://drive.google.com/file/d/1M7zJADX5-Z8LbWN_eUqNTLEP6urXIkgf/view?usp=sharing)" and "[Reveal ICO key](https://drive.google.com/file/d/1J9HfN5cFa7dtQX1PCA6udLO8K1FP3pkT/view?usp=sharing)" dialogs. The latter does not have the name of the wallet in the header, please check the screenshots on the previous links.

**Comment by danielmain** [comment]
@gabriela-ponce thank you for the review! About the inconsistencies. The original one is the "Wallet public Key", that is why the wallet name is present. I had no screens for the dialog and I intentionally removed the wallet name in the ICO dialog in order to **avoid confusion that this key is NOT the wallet-pub-key**. In case we should include it, I can do it. I can ask Darko tomorrow.

**Comment by dmitrii-gaico** [comment]
> @danielmain The fixes for [these](https://github.com/input-output-hk/daedalus/pull/2604#issuecomment-867876889) issues look good, great work 👍
> I found another small inconsistency between the "[Reveal public key](https://drive.google.com/file/d/1M7zJADX5-Z8LbWN_eUqNTLEP6urXIkgf/view?usp=sharing)" and "[Reveal ICO key](https://drive.google.com/file/d/1J9HfN5cFa7dtQX1PCA6udLO8K1FP3pkT/view?usp=sharing)" dialogs. The latter does not have the name of the wallet in the header, please check the screenshots on the previous links.

From my side I think we should display the wallet name for ICO as well as each Wallet has it's own ICO Key.
@darko-mijic Please advise about question above. Thank you in advance.

**Comment by darko-mijic** [comment]
> Please advise if we should display the wallet name in ICO Spending psw dialog. Thank you in advance.

Yes, please. Good catch!

**Comment by nikolaglumac** [comment]
Thanks @darko-mijic!

@dmitrii-gaico @gabriela-ponce please confirm that adding the wallet's name to the title section of the dialog is the last remaining issue to be fixed 🙏

**Comment by danielmain** [comment]
Wallet Name added in the last [commit](https://github.com/input-output-hk/daedalus/pull/2604/commits/a82b77f8770c36611d45372e284ccc4ff88b7686)

File: source/renderer/app/actions/wallets-actions.js [hunk 8]
```diff
@@ -78,7 +78,9 @@ export default class WalletsActions {
   }> = new Action();
   saveQRCodeImageSuccess: Action<{ walletAddress: string }> = new Action();
   getAccountPublicKey: Action<{ spendingPassword: string }> = new Action();
-  copyPublicKey: Action<{ publicKey: string }> = new Action();
+  getICOPublicKey: Action<{ spendingPassword: string }> = new Action();
+  copyWalletPublicKey: Action<{ publicKey: string }> = new Action();
```
Comment (nikolaglumac) [inline source/renderer/app/actions/wallets-actions.js:82]: Please rename `copyWalletPublicKey ` into `copyAccountPublicKey `

File: source/renderer/app/api/common/types.js [hunk 9]
```diff
@@ -6,3 +6,9 @@ export type RequestConfig = $Exact<{
   cert: Uint8Array,
   key: Uint8Array,
 }>;
+
+export type MessageDescriptor = {
```
Comment (nikolaglumac) [inline source/renderer/app/api/common/types.js:10]: Please use https://github.com/input-output-hk/daedalus/blob/a82b77f8770c36611d45372e284ccc4ff88b7686/source/renderer/app/types/i18nTypes.js#L2

File: source/renderer/app/components/wallet/settings/PublicKeyField.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+import styles from './PublicKeyField.scss';
+
+type Props = {
+  publicKey: string,
+  locale: Locale,
+  onCopyPublicKey: Function,
+  onShowQRCode: Function,
+  onOpenWalletKeyDialog: Function,
+  intl: intlShape.isRequired,
+  messages: Object,
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/settings/PublicKeyField.js:25]: Please add a proper type definition: `messages: { [string]: MessageDescriptor },`

**Comment by alexander-rukin** [comment]
image

- copy icon is too big -> https://zpl.io/aNrBD1n reference 
- qr - ok 
- eye (both: crossed and normal) shifted 1px to the top (move 1px bot pls)

**Comment by alexander-rukin** [comment]
image

distance after derivation path string is too big - reference https://zpl.io/VOZ5jAr -> it should be 11px (now looks like 30px) 
it's problem for share public key and multisig key.
for the receive screen it's bugged too, but in opposite way

File: source/renderer/app/actions/wallets-actions.js (hunk 8 above)
Comment (danielmain) [reply source/renderer/app/actions/wallets-actions.js:82]: This is going to be done in the scope of [DDW-709](https://jira.iohk.io/browse/DDW-709)

Comment (nikolaglumac) [reply source/renderer/app/actions/wallets-actions.js:82]: Perfect 👌

**Comment by danielmain** [comment]
- [ ] Code review suggestions applied (hope I could cover all your topics)
- [ ] Centered icons and replaced some svgs
- [ ] Added new strings en/jp Multi-signature

**Comment by alexander-rukin** [comment]
image

icon seems to be **missaligned** with a text + opacity is 0.5 here -> **should be 0.7**

https://zpl.io/VOZ5jAr - reference 

image

**Comment by alexander-rukin** [comment]
image

width of a text is calculated wierdly - why 171px? 
it's more logical to have 135px there:
28*3+11+10+10+20 = 135
– 28 - icon width 
– 11 - right margin 
– 10 - margin between 
– 20 - left margin

**Review by alexander-rukin — APPROVED** [review]
Looks good

**Review by dmitrii-gaico — APPROVED** [review]

**Comment by dmitrii-gaico** [comment]
Ready to go!

**Review by miorsufianiohk — APPROVED** [review]
LGTM 👍

**Review by gabriela-ponce — APPROVED** [review]

File: source/renderer/app/assets/images/copy.inline.svg [hunk 11]
```diff
@@ -1,15 +1,16 @@
...
-            <g>
-                <g>
-                    <g>
-                        <g>
-                            <path d="M16.005 6C17.107 6 18 6.89 18 7.991V26.01c0 1.1-.893 1.991-1.995 1.991H1.995C.893 28 0 27.11 0 26.009V7.99C0 6.891.893 6 1.995 6h14.01zm-2.007 3H4.002C3.456 9 3 9.448 3 10v14c0 .555.449 1 1.002 1h9.996c.546 0 1.002-.448 1.002-1V10c0-.555-.449-1-1.002-1zm8.007-9C23.107 0 24 .89 24 1.991V20.01c0 1.1-.893 1.991-1.995 1.991H20v-3c.507-.002.935-.387.993-.884l.007-.117V4.001C21 3.445 20.551 3 19.998 3h-9.996C9.456 3 9 3.448 9 4H6V1.991C6 .891 6.893 0 7.995 0h14.01z" transform="translate(-1462 -364) translate(60 264) translate(0 17) translate(1386 69) translate(16 14)"/>
-                        </g>
-                    </g>
-                </g>
+<svg viewBox="0 0 24 28" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+  <title>clipboard-ic</title>
```
Comment (nikolaglumac) [inline source/renderer/app/assets/images/copy.inline.svg:2]: @danielmain please remove this <title> line 🙏

File: source/renderer/app/assets/images/hide-key.inline.svg [hunk 12]
```diff
@@ -1,13 +1,16 @@
...
-    <g fill="none" fill-rule="evenodd">
-        <g fill="#5E6066">
-            <g>
-                <g>
-                    <g>
-                        <path d="M14 5.09c3.564 0 6.364 2.8 6.364 6.365 0 .763-.128 1.654-.51 2.29l3.691 3.691c1.91-1.654 3.437-3.69 4.328-5.981-2.164-5.6-7.637-9.546-14-9.546-1.782 0-3.437.382-5.091.891l2.8 2.8c.763-.382 1.654-.51 2.418-.51zM1.273 1.656L4.2 4.582l.51.509C2.544 6.745.89 8.909 0 11.455 2.164 17.055 7.636 21 14 21c1.91 0 3.818-.382 5.6-1.018l.51.509 3.69 3.69 1.655-1.654L2.927 0 1.273 1.655zm7 7l1.909 1.909v.89c0 2.164 1.654 3.819 3.818 3.819.255 0 .51 0 .89-.128l1.91 1.91c-.89.509-1.782.763-2.8.763-3.564 0-6.364-2.8-6.364-6.363 0-1.019.255-1.91.637-2.8zm5.472-1.019l3.946 3.946v-.255c0-2.163-1.655-3.818-3.818-3.818 0 .127 0 .127-.128.127z" transform="translate(-1636 -366) translate(60 264) translate(0 17) translate(1562 69) translate(14 16)"/>
-                    </g>
-                </g>
+<svg width="28px" height="24px" viewBox="0 0 28 24" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+  <title>hide</title>
```
Comment (nikolaglumac) [inline source/renderer/app/assets/images/hide-key.inline.svg:2]: @danielmain please remove this <title> line 🙏

File: source/renderer/app/assets/images/qr-code.inline.svg [hunk 13]
```diff
@@ -1,3 +1,16 @@
-<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24">
-    <path fill="#5E6066" fill-rule="evenodd" d="M24 16v4h-7v-3h-2v7h-2V13h6v3h3v-3h2v3zm-7 6h2v2h-2v-2zm4 0h2v2h-2v-2zM2 2v7h7V2H2zM0 0h11v11H0V0zm2 15v7h7v-7H2zm-2-2h11v11H0V13zM15 2v7h7V2h-7zm-2-2h11v11H13V0zM4 4h3v3H4V4zm0 13h3v3H4v-3zM17 4h3v3h-3V4z"/>
+<svg width="28px" height="28px" viewBox="0 0 28 28" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+  <title>qr</title>
```
Comment (nikolaglumac) [inline source/renderer/app/assets/images/qr-code.inline.svg:2]: @danielmain please remove this <title> line 🙏

File: source/renderer/app/assets/images/reveal-key.inline.svg [hunk 14]
```diff
@@ -1,13 +1,16 @@
...
-    <g fill="none" fill-rule="evenodd">
-        <g fill="#5E6066">
-            <g>
-                <g>
-                    <g>
-                        <path d="M14 0C7.636 0 2.164 3.945 0 9.545c2.164 5.6 7.636 9.546 14 9.546s11.836-3.946 14-9.546C25.836 3.945 20.364 0 14 0zm0 15.91c-3.564 0-6.364-2.8-6.364-6.365 0-3.563 2.8-6.363 6.364-6.363s6.364 2.8 6.364 6.363c0 3.564-2.8 6.364-6.364 6.364zm0-10.183c-2.164 0-3.818 1.655-3.818 3.818 0 2.164 1.654 3.819 3.818 3.819s3.818-1.655 3.818-3.819c0-2.163-1.654-3.818-3.818-3.818z" transform="translate(-1636 -165) translate(60 60) translate(0 64) translate(1576 41)"/>
-                    </g>
-                </g>
+<svg width="28px" height="20px" viewBox="0 0 28 20" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+  <title>reveal</title>
```
Comment (nikolaglumac) [inline source/renderer/app/assets/images/reveal-key.inline.svg:2]: @danielmain please remove this <title> line 🙏

**Review by nikolaglumac — CHANGES_REQUESTED** [review]
@danielmain please add the requested changes 🙏

**Review by daniloprates — APPROVED** [review]
Amazing work @danielmain 💯
